rioxarray/rioxarray.py

```diff
@@ -327,6 +332,32 @@ def _int_bounds(self):
         bottom = float(self._obj[self.y_dim][-1])
         return left, bottom, right, top
 
+    def _check_dimensions(self):
+        """
+        This function validates that the dimensions 2D/3D and
+        they are are in the proper order.
+
+        Returns:
+        --------
+        str or None: Name extra dimension.
+        """
+        extra_dims = list(set(list(self._obj.dims)) - set([self.x_dim, self.y_dim]))
+        if len(extra_dims) > 1:
+            raise TooManyDimensions("Only 2D and 3D data arrays supported.")
+        elif extra_dims and self._obj.dims != (extra_dims[0], self.y_dim, self.x_dim):
```

Comment:
This check seems odd to me since it depends on the order of `extra_dims`. If the data is 4D (which is not checked here) then the "bands" dimension could be anywhere. Also since you converted the dims to sets and then back to lists you don't know the order of the extra dims.

Do you have anything that defines the name of the "bands" dimension? Sorry, not super familiar with rioxarray.

Reply:
You could use xarray's `transpose` method to shift dimensions around too.

Reply:
> This check seems odd to me since it depends on the order of extra_dims. 

Currently it is okay since the dataset is ensured to have only 2 or 3 dimensions. So, there should only be up to 1 extra dimension.

> If the data is 4D (which is not checked here) then the "bands" dimension could be anywhere. 

Currently 4D data is not supported or allowed. But, it would be nice to eventually reach n-dim support.

> Do you have anything that defines the name of the "bands" dimension?

If the dataset is 2D, then there won't be a bands dimension and it is just 1 band. But, if the dataset is 3D, then it will always be the dimension that is not the `x` or `y` dimension.

Reply:
> You could use xarray's transpose method to shift dimensions around too.

That is a really nice option. I have been thinking about how to handle that as I am not sure how I feel about it being magically handled. For example, when reading the dataset back in from a tif using `open_rasterio`, the dimension order would be swapped and the user wouldn't know it. Maybe recommending swapping the dimension order using `DataArray.transpose` in the error message would be useful. 

Thoughts?

Reply:
Yeah, this library is lower level than `trollimage` which is where most of my rasterio writing experience comes from. I think in this case less magic behavior is probably better.

As for 2D and 3D being the only ones allowed, is that check done outside of this method? The top if statement in this PR only checks for more than 1 dimension I thought.

Reply:
🤦‍♂ Ah you are checking that there is only 1 extra dimension. Got it.

Reply:
```python
extra_dims = list(set(list(self._obj.dims)) - set([self.x_dim, self.y_dim]))
```
In this case, the extra_dims are the dimensions not in the expected `x` and `y` dimensions. So, if it does not have an extra dimension, it is 2D. If it has 1 extra dimension it is 3D, etc ...